src/Libraries/Microsoft.Extensions.Http.Resilience/Internal/RequestMessageSnapshot.cs

```diff
@@ -64,24 +86,76 @@ bool IResettable.TryReset()
         _method = null;
         _version = null;
         _requestUri = null;
+        if (_content is StreamContent)
+        {
+            // a snapshot's StreamContent is always a unique copy (deep clone)
+            // therefore, it is safe to dispose when snapshot is no longer needed
+            _content.Dispose();
+        }
+
         _content = null;
 
         return true;
     }
 
     void IDisposable.Dispose() => _snapshots.Return(this);
 
-    private void Initialize(HttpRequestMessage request)
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Resilience", "EA0014:The async method doesn't support cancellation", Justification = "Past the point of no cancellation.")]
+    private static async Task<(HttpContent? content, HttpContent? clonedContent)> CloneContentAsync(HttpContent? content)
     {
-        if (request.Content is StreamContent)
+        HttpContent? clonedContent = null;
+        if (content is not null)
         {
-            Throw.InvalidOperationException($"{nameof(StreamContent)} content cannot by cloned.");
+            HttpContent originalContent = content;
+            Stream originalRequestBody = await content.ReadAsStreamAsync().ConfigureAwait(false);
+            MemoryStream clonedRequestBody = new MemoryStream();
+            await originalRequestBody.CopyToAsync(clonedRequestBody).ConfigureAwait(false);
```

Comment:
The streamed content can point to a large file which can have > GBs in size. Since we are copying to memory we should constraint it. Let's say < 10MB

Reply:
I'm not sure if it's a good idea to impose a size limit. For example, I don't think the framework imposes any such limit when cloning request content for redirected requests. IMO, we should allow this to be limitless, and require users to enforce their own request content size limits, if they have them, via server configurations and/or their own custom handlers, filters, middleware, etc. Thoughts?

Reply:
This change can buffer potentially endless stream into memory and crash the process. I think it's something we should guard against.

This thing should be used for relatively small streamed payloads.

Btw, you can try to call `LoadIntoBufferAsync` on the content to see how it behaves.

Reply:
I do understand the concern; however, the max size that a given stream can grow before it crashes the process is bound only by the memory resources available on the given server that it's executing on. Therefore, by choosing a static max limit, we are creating a cap that might unnecessarily hinder users of the API that have both requirements and the necessary server resources available to process requests with exceptionally large stream content. Further, `RequestMessageSnapshot` is only cloning content from _existing_ request objects that contain _existing_ `StreamContent`. Therefore, a `HttpRequestMessage` object has already been created with `StreamContent` of the given size before any of the `RequestMessageSnapshot` operations are executed. If the size of the `StreamContent` was in fact too large, it would have already crashed the process.

I feel strongly that we should not impose a size limit when cloning the StreamContent.

Reply:
> If the size of the StreamContent was in fact too large, it would have already crashed the process.

This is not true, the `StreamContent` can point to file stream and it consumes minimal memory even as you read the whole file. What you are doing here is materializing the whole stream into the memory. This is what concerns me.

Reply:
The only way what you are saying is true is if the content is a `FileStream` and is read into a buffer, correct? This is somewhat of a remarkable edge case since we are talking about streams in the context of HTTP request content. It would help me understand this better if you could provide an example scenario where a developer would want to do this.

That said, the idea of supporting buffered content is certainly not outside the realm of possibilities given that this will be an extension method exposed by the framework itself and we should do our best to cover all possible use cases. I still don't agree that we should enforce any kind of size limit; however, I do believe that we should add support for utilizing a buffer if the given request's `StreamContent` supports it. To this end, I responded to your other comment proposing support for utilizing a given `StreamContent`'s buffer if said `StreamContent` permits it to be read from.